service/src/main/java/bio/terra/catalog/service/dataset/DatasetDao.java

```diff
@@ -112,15 +115,33 @@ public Dataset mapRow(ResultSet rs, int rowNum) throws SQLException {
   }
 
   @ReadTransaction
-  public List<Dataset> find(StorageSystem storageSystem, Collection<String> ids) {
-    if (ids.isEmpty()) {
+  // This code is safe because it builds a template query string using ?s only. It relies on
+  // JdbcTemplate to perform all text substitutions.
+  @SuppressWarnings("java:S2077")
+  public List<Dataset> find(Map<StorageSystem, Collection<String>> systemsAndIds) {
+    String query = "(storage_system = ? AND storage_source_id IN (%s))";
+    List<Object> args = new ArrayList<>();
+    String whereClause =
+        systemsAndIds.entrySet().stream()
+            .filter(entry -> !entry.getValue().isEmpty())
+            .map(
+                entry -> {
+                  args.add(String.valueOf(entry.getKey()));
+                  args.addAll(entry.getValue());
```

Comment:
nit: Is there any way we could cleanly separate the arg building from the map? The rationale being the map function has a side effect of building the arguments.

I was thinking of something like having the map return tuples of args and clauses, but I think that would add too much complexity...

I know you mentioned you weren't happy with the query building, and this isn't critical.

Reply:
I thought about this; having one method that builds the query string and another that populates the args map. The main problems with doing it separately is that you end up duplicating the the `stream()` operation and it's less clear (to me at least) that the query string is mirroring the arg list. On the other hand, doing it separately means that it's more obvious that the query only contains `?`s and not any other data.

I did look a bit but didn't see anything that would help here, other than using a higher level query language to write our DAO in. Spring supports a few ways to do this, but that would be a pretty big change to the DAO code. It might be worthwhile looking into at some point though, as our DAO is pretty simple so far.